airbyte-workers/src/main/java/io/airbyte/workers/internal/EmptyAirbyteSource.java

```diff
@@ -4,30 +4,90 @@
 
 package io.airbyte.workers.internal;
 
+import com.fasterxml.jackson.databind.JsonNode;
 import io.airbyte.commons.json.Jsons;
+import io.airbyte.config.ResetSourceConfiguration;
+import io.airbyte.config.StreamDescriptor;
 import io.airbyte.config.WorkerSourceConfig;
+import io.airbyte.protocol.models.AirbyteGlobalState;
 import io.airbyte.protocol.models.AirbyteMessage;
 import io.airbyte.protocol.models.AirbyteMessage.Type;
 import io.airbyte.protocol.models.AirbyteStateMessage;
+import io.airbyte.protocol.models.AirbyteStateMessage.AirbyteStateType;
+import io.airbyte.protocol.models.AirbyteStreamState;
+import io.airbyte.protocol.models.StateMessageHelper;
+import io.vavr.control.Either;
 import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.EmptyStackException;
+import java.util.LinkedList;
+import java.util.List;
 import java.util.Optional;
+import java.util.Queue;
+import java.util.Stack;
 import java.util.concurrent.atomic.AtomicBoolean;
+import lombok.extern.slf4j.Slf4j;
 
 /**
  * This source will never emit any messages. It can be used in cases where that is helpful (hint:
  * reset connection jobs).
  */
+@Slf4j
 public class EmptyAirbyteSource implements AirbyteSource {
 
   private final AtomicBoolean hasEmittedState;
+  private final Queue<StreamDescriptor> streamDescriptors = new LinkedList<>();
+  private boolean isPartialReset;
+  private boolean isStarted = false;
+  private Either<List<AirbyteStateMessage>, AirbyteStateMessage> perStreamOrGlobalState;
 
   public EmptyAirbyteSource() {
     hasEmittedState = new AtomicBoolean();
   }
 
   @Override
   public void start(final WorkerSourceConfig sourceConfig, final Path jobRoot) throws Exception {
-    // no op.
+
+    try {
+      if (sourceConfig == null || sourceConfig.getSourceConnectionConfiguration() == null) {
+        isPartialReset = false;
+      } else {
+        ResetSourceConfiguration sourceConfiguration = Jsons.object(sourceConfig.getSourceConnectionConfiguration(), ResetSourceConfiguration.class);
+        streamDescriptors.addAll(sourceConfiguration.getStreamsToReset());
+        if (streamDescriptors.isEmpty()) {
+          isPartialReset = false;
+        } else {
+          Either<JsonNode, List<AirbyteStateMessage>> eitherState = StateMessageHelper.getTypedState(sourceConfig.getState().getState());
+          if (eitherState.isLeft()) {
+            log.error("The state is not compatible with a partial reset that  have been requested");
+            throw new IllegalStateException("Legacy state for a partial reset");
+          }
```

Comment:
I'm confused by this. I was under the impression that we will always be persisting `streamsToReset` in the ResetSourceConfiguration, even in the case where a connection still uses legacy state (it would just be persisting _all_ streams in the connection in that case).

But looking at the logic in this method, it seems that you expect the sourceConnectionConfiguration to always be `null` if the connection uses legacy state. And that would mean we need to conditionally change how we set the sourceConnectionConfiguration in the temporal workflow based on the connection's current state type, which I wasn't aware of.

Reply:
I have updated it with a check that compare the stream in the catalog and in the stream. If we have a legacy state, they have to be the same.